pkg/rpc/breaker_new_probe.go

```diff
@@ -0,0 +1,357 @@
+		// (*Connection).ConnectNoBreaker, and they need to be signaled as well
+		// but aren't listening to the stopper.
+		p.mu.c.gatedCC.Resolve(nil, errQuiescing)
+		done()
+	}
+}
+
+func (p *breakerProbe) run(ctx context.Context, report func(error), done func()) {
+	var t timeutil.Timer
+	defer t.Stop()
+	defer done()
+	// Immediately run probe after breaker circuit is tripped, optimizing for the
+	// case in which we can immediately reconnect.
+	t.Reset(0)
+	for {
+		if p.snap().deleted {
+			return
+		}
+
+		// NB: we don't need to close initialHeartbeatDone in these error cases.
+		// Connect() is cancellation-sensitive as well.
+		select {
+		case <-ctx.Done():
+			// Stopper quiescing, node shutting down. Mirroring what breakerProbe.launch
+			// does when it can't launch an async task: leave the broken connection around,
+			// no need to close initialHeartbeatDone, just report errQuiescing and quit.
+			report(errQuiescing)
+			return
+		case <-t.C:
+			t.Read = true
+			// Retry every second. Note that if runHeartbeatUntilFailure takes >1, we'll
+			// retry immediately once it returns. This means that a connection breaking
+			// for the first time is usually followed by an immediate redial attempt.
+			t.Reset(p.heartbeatInterval)
+		}
+
+		// Peer is currently initializing (first use) or unhealthy (looped around
+		// from earlier attempt). `runOnce` will try to establish a connection and
+		// keep it healthy for as long as possible. On first error, it will return
+		// back to us.
+		err := p.runOnce(ctx, report)
+		// If ctx is done, Stopper is draining. Unconditionally override the error
+		// to clean up the logging in this case.
+		if ctx.Err() != nil {
+			err = errQuiescing
+		}
+
+		// Transition peer into unhealthy state.
+		now := p.opts.Clock.Now()
+		p.onHeartbeatFailed(ctx, err, now, report)
+
+		// Release peer and delete from map, if appropriate. We'll detect
+		// whether this happened after looping around.
+		p.maybeDelete(now)
+
+		if errors.Is(err, errQuiescing) {
+			// Heartbeat loop ended due to shutdown. Exit the probe, it won't be
+			// started again since that means running an async task through the
+			// Stopper.
+			return
+		}
+
+		p.mu.Lock()
+		p.mu.c = newConnectionToNodeID(p.k, p.mu.c.breakerSignalFn)
+		p.mu.Unlock()
+	}
+}
+
+func (p *breakerProbe) onHeartbeatFailed(
+	ctx context.Context, err error, now time.Time, report func(err error),
+) {
+	prevErr := p.b.Signal().Err()
+	// For simplicity, we have the convention that this method always returns
+	// with an error. This is easier to reason about since we're the probe,
+	// and - morally speaking - the connection is healthy as long as the
+	// probe is running and happy. We don't want to consider a connection
+	// healthy when the probe is not running but didn't report an error.
+	if err == nil {
+		err = errors.AssertionFailedf("unexpected connection shutdown")
+	}
+
+	touchOldPeers(p.peers, now)
```

Comment:
Why do we touch all peers here? How does failure of this connection link us to other potential peers?

Reply:
It's just a convenient trigger, added this comment (with next push)

	// There might be other peers in the map that are pending deletion, but which
	// are no longer seeing activity. To eventually clear them out, we check all
	// conns when any conn fails. This avoids the need to have an extra goroutine
	// sitting in all of the goroutine stacks we look at during support.
	touchOldPeers(p.peers, now)